python/pyarrow/tests/test_dataset.py

```diff
@@ -0,0 +1,350 @@
+def dataset(simple_data_source, tree_data_source, schema):
+    return ds.Dataset([simple_data_source, tree_data_source], schema)
+
+
+def test_scan_context():
+    context = ds.ScanContext()
+    assert isinstance(context.memory_pool, pa.MemoryPool)
+    context = ds.ScanContext(pa.default_memory_pool())
+    assert isinstance(context.memory_pool, pa.MemoryPool)
+
+
+def test_simple_data_source(record_batch):
+    source = ds.SimpleDataSource([record_batch])
+    assert isinstance(source, ds.SimpleDataSource)
+    assert source.partition_expression is None
+
+
+def test_tree_data_source(simple_data_source):
+    source = ds.TreeDataSource([simple_data_source] * 2)
+    assert isinstance(source, ds.TreeDataSource)
+
+
+def test_filesystem_data_source(mockfs):
+    file_format = ds.ParquetFileFormat()
+    source_partition = None
+    path_partitions = {}
+
+    paths = [
+        'subdir/1/xxx/file0.parquet',
+        'subdir/2/yyy/file1.parquet',
+    ]
+    source = ds.FileSystemDataSource(
+        mockfs,
+        paths,
+        source_partition=source_partition,
+        path_partitions=path_partitions,
+        file_format=file_format
+    )
+
+    source_partition = ds.ComparisonExpression(
+        ds.CompareOperator.Equal,
+        ds.FieldExpression('source'),
+        ds.ScalarExpression(1337)
+    )
+    path_partitions = {
+        'subdir/1/xxx/file0.parquet': ds.ComparisonExpression(
+            ds.CompareOperator.Equal,
+            ds.FieldExpression('part'),
+            ds.ScalarExpression(1)
+        ),
+        'subdir/2/xxx/file1.parquet': ds.ComparisonExpression(
+            ds.CompareOperator.Equal,
+            ds.FieldExpression('part'),
+            ds.ScalarExpression(2)
+        )
+    }
+    source = ds.FileSystemDataSource(
+        mockfs,
+        paths,
+        source_partition=source_partition,
+        path_partitions=path_partitions,
+        file_format=file_format
+    )
+    assert source.partition_expression.equals(source_partition)
+
+
+def test_dataset(simple_data_source, tree_data_source, schema):
+    dataset = ds.Dataset([simple_data_source, tree_data_source], schema)
+
+    assert isinstance(dataset, ds.Dataset)
+    assert isinstance(dataset.schema, pa.Schema)
+    for source in dataset.sources:
+        assert isinstance(source, ds.DataSource)
+
+    condition = ds.ComparisonExpression(
+        ds.CompareOperator.Equal,
+        ds.FieldExpression('i64'),
+        ds.ScalarExpression(1)
+    )
+    # TODO(kszucs): test non-boolean expressions for filter do raise
+    builder = dataset.new_scan().use_threads(True).filter(condition)
+    assert isinstance(builder, ds.ScannerBuilder)
+    assert isinstance(builder.schema, pa.Schema)
+
+    scanner = builder.finish()
+    assert isinstance(scanner, ds.Scanner)
+
+    for task in scanner.scan():
+        assert isinstance(task, ds.ScanTask)
+        for record_batch in task.scan():
+            assert isinstance(record_batch, pa.RecordBatch)
```

Comment:
Also assert the record batch's contents/row count; should have rows containing only `[1, 1.0]` IIUC

Reply:
Added.